src/bin/bobd.rs

```diff
@@ -1,23 +1,61 @@
-use bob::api::grpc::{server, PutRequest, GetRequest, OpStatus, BlobResult};
+use bob::api::grpc::{server, PutRequest, GetRequest, OpStatus, BlobResult, Null};
 
 use futures::{future, Future, Stream};
+use futures::future::ok;
 use tokio::executor::DefaultExecutor;
 use tokio::net::TcpListener;
 use tower_h2::Server;
 use tower_grpc::{Request, Response};
+use bob::core::grinder::{Grinder};
+use bob::core::data::{BobKey, BobData, BobOptions};
+use bob::core::backend::Backend;
+use bob::core::sprinkler::{Sprinkler};
+use tokio::runtime::{Runtime};
+extern crate clap;
 
-#[derive(Clone, Debug)]
-struct BobSrv;
+use clap::{App, Arg};
+
+#[derive(Clone)]
+struct BobSrv {
+    grinder: Grinder
+}
+
+impl BobSrv {
+    pub fn get_periodic_tasks(&self, ex: tokio::runtime::TaskExecutor) -> Box<impl Future<Item=(), Error=()>> {
+        self.grinder.get_periodic_tasks(ex)
+    }
+}
 
 impl server::BobApi for BobSrv {
-        type PutFuture = future::FutureResult<Response<OpStatus>, tower_grpc::Status>;
+        type PutFuture = Box<Future<Item=Response<OpStatus>,Error=tower_grpc::Status> + Send>;
         type GetFuture = future::FutureResult<Response<BlobResult>, tower_grpc::Status>;
+        type PingFuture = future::FutureResult<Response<Null>, tower_grpc::Status>;
 
-        fn put(&mut self, _request: Request<PutRequest>) -> Self::PutFuture {
-            let response = Response::new(OpStatus {
+        fn put(&mut self, req: Request<PutRequest>) -> Self::PutFuture {
+            let param = req.get_ref();
+            let f = self.grinder.put(
+                BobKey{
+                    key: param.key.as_ref().unwrap().key.clone()
+                    },
+                BobData{
+                    data: param.data.as_ref().unwrap().data.clone()
```

Comment:
If I understand correctly, you are creating a copy of the array with the data. This array can be large. Is there a way to transfer array of data without cloning?

Reply:
@pyakushin  do we need in structure BobData field data declare via Arc? Would be this enought?

Reply:
You can get ownership of `PutRequest`,  with [```Request::into_inner()```](https://tower-rs.github.io/tower-grpc/tower_grpc/struct.Request.html#method.into_inner). `PutRequest` has fields type of `Option<_>`, you can take value from it with [```Option::take()```](https://doc.rust-lang.org/std/option/enum.Option.html#method.take)